apps/designer/app/shared/copy-paste/use-instance-copy-paste.ts

```diff
@@ -0,0 +1,163 @@
+import { utils } from "@webstudio-is/project";
+import type { Instance, UserProp } from "@webstudio-is/react-sdk";
+import { useEffect, useMemo } from "react";
+import { type InstanceCopyData, serialize, deserialize } from "./serialize";
+
+const isInstanceClipboardEvent = (
+  event: ClipboardEvent,
+  options: { allowAnyTarget?: boolean }
+) => {
+  const selection = document.getSelection();
+  if (selection !== null && selection.type === "Range") {
+    return false;
+  }
+
+  // Note on event.target:
+  //
+  //   The spec (https://w3c.github.io/clipboard-apis/#to-fire-a-clipboard-event)
+  //   says that if the context is not editable, the target should be the focused node.
+  //
+  //   But in practice it seems that the target is based
+  //   on where the cursor is, rather than which element has focus.
+  //   For example, if a <button> has focus, the target is the <body> element
+  //   (at least in Chrome).
+
+  // If cursor is in input,
+  // don't copy instance (we may want to add more exceptions here in the future)
+  if (
+    options.allowAnyTarget !== true &&
+    (event.target instanceof HTMLInputElement ||
+      event.target instanceof HTMLTextAreaElement)
+  ) {
+    return false;
+  }
+
+  return true;
+};
+
+type Props = {
+  selectedInstanceData?: InstanceCopyData | undefined;
+  allowAnyTarget?: boolean;
+  onPaste?: (instance: Instance, props?: UserProp[]) => void;
+  onCut?: (instance: Instance) => void;
```

Comment:
these are not really optional, right?

Reply:
They are, but no sure why I did that. Will fix.